### PR TITLE
Butterworth filtering function for AnalogSignal

### PR DESCRIPTION
**Pull request by Junji110**
Added a Butterworth filtering function `butter()` in a newly created module `signal`

### CODE REVIEW

**Comment by Junji110** [comment]
I had used matplotlib.mlab.psd() in the unit tests, but I noticed that elephant is not dependent on matplotlib and so this function shouldn't be used.
In the latest commit I replaced all occurrences of matplotlib.mlab.psd() with an (almost) equivalent function in scipy, i.e. scipy.signal.welch().

**Comment by Junji110** [comment]
according to the discussion in PR #22, the module name was changed.

File: elephant/signal_processing.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+def butter(anasig, highpassfreq=None, lowpassfreq=None, order=4,
+           filtfunc='filtfilt'):
+    """
+    Butterworth filtering function for AnalogSignal objects. Filter type is
+    determined according to how values of highpassfreq and lowpassfreq are
+    given (see Parameters for details).
+
+    Parameters
+    ----------
+    anasig: AnalogSignal
```
Comment (apdavison) [inline elephant/signal_processing.py]: use `signal` intead of `anasig`.

File: elephant/signal_processing.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        time series data to be filtered
+    highpassfreq, lowpassfreq: Quantity or None
```
Comment (apdavison) [inline elephant/signal_processing.py]: I think one or more underscores in these argument names would help readability.

File: elephant/signal_processing.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        high-pass and low-pass cut-off frequencies, respectively.
+        Filter type is determined depending on values of these arguments:
+            highpassfreq only (lowpassfreq = None):    highpass filter
+            lowpassfreq only (highpassfreq = None):    lowpass filter
+            highpassfreq < lowpassfreq:    bandpass filter
+            highpassfreq > lowpassfreq:    bandstop filter
+    order: int
+        Order of Butterworth filter. Default is 4.
+    filtfunc: string
```
Comment (apdavison) [inline elephant/signal_processing.py]: `filter_function`

File: elephant/signal_processing.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        Filtering function to be used. Either 'filtfilt'
+        (scipy.signal.filtfilt) or 'lfilter' (scipy.signal.lfilter). In most
+        applications 'filtfilt' should be used, because it doesn't bring about
+        phase shift. Default is 'filtfilt'.
+
+    Returns
+    -------
+    anasig_out: AnalogSignal
```
Comment (apdavison) [inline elephant/signal_processing.py]: `filtered_signal` or `signal_out`

File: elephant/signal_processing.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        Filtered signal
+    """
+
+    def apply_filter(signal, Fs, highpassfreq=None, lowpassfreq=None, order=4,
```
Comment (apdavison) [inline elephant/signal_processing.py]: Is this a public function? If so it needs a docstring (docstring would be good anyway). If it is an internal function, the name should start with an underscore.

File: elephant/signal_processing.py (hunk 1 above)
Comment (apdavison) [inline elephant/signal_processing.py]: Also, as noted for the other open pull requests, the function should operate on `AnalogSignalArray`.

**Comment by apdavison** [comment]
By the way, my apologies if my comments seem rather brief and hence peremptory. It is just that I am typing with one hand this week following a ski accident so I'm avoiding too much typing! Of course all comments are just suggestions and are up for discussion!

**Comment by Junji110** [comment]
i've made fixes according to the comments.
I've also made changes to properly deal with inputs given as Quantity array or NumPy ndarray.
Please review the code again.

File: elephant/signal_processing.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                btype = 'bandstop'
+        elif lpfreq:
+            Wn = lpfreq / Fn
+            btype = 'lowpass'
+        elif hpfreq:
+            Wn = hpfreq / Fn
+            btype = 'highpass'
+        else:
+            raise ValueError(
+                "Either highpass_freq or lowpath_freq must be given"
```
Comment (dholstein) [inline elephant/signal_processing.py]: Little typo: path -> pass

**Pull request merged**